src/libopensc/card-entersafe.c

```diff
@@ -345,14 +378,12 @@ static int entersafe_mac_apdu(sc_card_t *card, sc_apdu_t *apdu,
 	 apdu->datalen=apdu->lc;
 
 out:
-	 if(tmp)
-		  free(tmp);
-	 if(tmp_rounded)
-		  free(tmp_rounded);
-	 if (ctx) {
-		sc_evp_cipher_free(alg);
-		EVP_CIPHER_CTX_free(ctx);
-	}
+	if (r != SC_SUCCESS)
+		sc_log_openssl(card->ctx);
```

Comment:
Calling the log function on the particular line where some of the openssl functions failed instead of doing it after the `goto out` has the advantage that you will see in the error log the file and line on which this macro was called, which simplifies debugging. When calling this from the `goto out`, one needs to guess from the error message which function failed so I would prefer to call this function before the `goto out` for example in this case.

Reply:
Okay, I'll break it down to the individual cases.